cabal-install/Distribution/Client/ProjectConfig.hs

```diff
@@ -567,15 +578,50 @@ data BadPackageLocationMatch
   deriving Show
 
 renderBadPackageLocations :: BadPackageLocations -> String
-renderBadPackageLocations (BadPackageLocations bpls) =
-    unlines (map renderBadPackageLocation bpls)
+renderBadPackageLocations (BadPackageLocations prov bpls)
+      -- There is no provenance information,
+      -- render standard bad package error information.
+    | null provenance = renderErrors renderBadPackageLocation
+
+      -- The configuration is implicit, render bad package locations
+      -- using possibly specialized error messages.
+    | provenance == [Implicit] = renderErrors renderImplicitBadPackageLocation
+
+      -- The configuration was read from one or more explicit path(s),
+      -- list the locations and render the bad package error information.
+      -- The intent is to supersede this with the relevant location information
+      -- per package error.
+    | otherwise =
+           "When using configuration(s) from "
+        ++ intercalate ", " (mapMaybe getExplicit provenance)
```

Comment:
Maybe this should also mention if the implicit is used? (I think this case is impossible, so outputting something in that case could help debugging in the future. Or maybe even an error.)